### PR TITLE
Improved security logging #664

### LINKED ISSUE #664: Improve security logging for rejected requests by spring firewall
If a request is rejected, we want a specific SECURITY log entry about

- IP adress 
- HTTP RQUEST information (if not already logged by spring security)

we use the existing security log service

### LINKED ISSUE #668: Change security log service, so providing always http request information + change format + change to warn
## Situation
We currently have not the client IP inside security logs and code parts identifying security risks would have to fetch http servlet request etc. every time, fetch data etc.

Also the output format is a little bit "bad readable" : [SECURITY] [USER:$user] (TYPE) : $customMessage
For example:
```
2021-06-23 10:37:57.401  INFO 7432 --- [nio-8443-exec-8] c.d.s.s.logging.SecurityLogService       : [SECURITY] [USER:null] (POTENTIAL INTRUSION) :Rejected request, ...
```

Also INFO is not a good level for security logs

## Wanted

- security log entries has to be log level: WARNING
- automatically fetch and provide clientIP and uri (relative path) of the current http request (if available)
- output format will change, it will contain the meta information (e.g. clientIP) as JSON output, not available or empty parts are simply not inside JSON:

 ``` 
2021-06-24 15:30:01.484  WARN 14829 --- [nio-8443-exec-8] c.d.s.s.logging.SecurityLogService       : [SECURITY] [POTENTIAL INTRUSION]:
data=
{
  "type" : "POTENTIAL_INTRUSION",
  "clientIp" : "127.0.0.1",
  "message" : "Rejected request, reason:{}",
  "messageParameters" : [ "The request was rejected because the HTTP method \"bad_request\" was not included within the list of allowed HTTP methods [HEAD, DELETE, POST, GET, OPTIONS, PATCH, PUT]" ],
  "requestURI" : "/api/anonymous/check/alive",
  "httpHeaders" : {
    "accept" : "*/*",
    "content-type" : "application/json",
    "host" : "localhost:8443",
    "user-agent" : "curl/7.68.0"
  },
  "sessionId" : "7DACF3AB4A756B6C0F45B3D40BEB6FA8"
}
, message=Rejected request, reason:The request was rejected because the HTTP method "bad_request" was not included within the list of allowed HTTP methods [HEAD, DELETE, POST, GET, OPTIONS, PATCH, PUT]
```
The message is at the end, so even when somebody calls security log service in wrong way (wrong amount of log parameters etc), the data will always appear as expected inside the logs.

## Here be dragons...
Inside the HTTP headers we have also contained the user credentials! 
Those credentials may never be logged in plain text, because we do not know if it is from an attacker of just a wrong call by a real user! 

We do following:
- Instead of logging plain text passwords, we will obfuscate them
- Obfuscated passwords having a minimum size will contain the first two characters. So we can check if password changes, and administrators can support customers more easy where a new requested API token is not used inside automated builds and build server call does always return a 403..
- The obfuscated part will also contain the used password length.
  - Example 1
    ```obfuscate("Basic aW350LXalc3RexampleXWzZXI6aW50LXRlc3Rfb23seXVzZXtHdk",52); ```  
     will return
    ```  "Basic aW***length:58```

  - Example 2   
    ```obfuscate("aW350LXalc3RexampleXWzZXI6aW50LXRlc3Rfb23seXVzZXtHdk",52);```  
     will return
    ```aW***length:52```

  - Example 3
     ```obfuscate("aW350LXalc3RxampleXWzZXI6aW50LXRlc3Rfb23seXVzZXtHdk",52);```  
     will return  
     ```***length:51```

### LINKED ISSUE #670: Log wrong usage by security log service
## Situation
Currently we do not log explicit by security log service:

- wrong usage of REST API
  - wrong HTTP method (e.g. POST instead GET)
  - wrong/not existing pathes (404)
- unauthenticated access (403)

## Wanted
- log those events also by security log service
- introduce WRONG_USAGE security log type

### LINKED ISSUE #672: gradle task `integrationtest`does not execute Junit5 tests
## Situation
Inside `sechub-integrationtest` we have a special task called `integrationtest` which
- starts integration test instances (PDS, SecHub server)
- executes tests
- stops integration test instances

Some tests are written as Junit5 tests (only scenario tests need the special junit4 rule but some other tests are already written as Junit5)

All other Junit5 tests inside the other gradle submodules are executed by gradle!
But our specialized task does only execute Junit4 tests!

## Wanted
`integrationtest` task shall execute Junit5 tests as well

### PR DESCRIPTION
**Pull request by de-jcup**
This PR 
- closes #664 
- closes #668
 - closes #670
 - closes #672

### CODE REVIEW

**Review by Jeeppler — CHANGES_REQUESTED** [review]
See comments.

File: sechub-integrationtest/send_curl_request.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash 
+
+# Usage: bad_request <url><verb>
+
+function usage(){
+    echo "usage: send_curl_request <url> <request method>"
+}
+
+if [ -z "$1" ] ; then
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-integrationtest/send_curl_request.sh:9]: done

File: sechub-integrationtest/send_curl_request.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+fi
+
+if [ -z "$2" ] ; then
+    echo "request method is missing as second parameter"
+    usage
+    exit 1
+fi
+
+
+curl --insecure -X $2 -H "Content-Type: application/json" $1
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-integrationtest/send_curl_request.sh:22]: done

File: sechub-integrationtest/src/main/java/com/daimler/sechub/integrationtest/SecurityTestHelper.java [hunk 3]
```diff
@@ -304,10 +304,29 @@ private String collectionOfSupportedCiphers() throws Exception {
...
+        List<String> commands = new ArrayList<>();
+        commands.add("./send_curl_request.sh");
+        commands.add(url);
+        commands.add(customRequestmethod);
+
+        ProcessBuilder pb = new ProcessBuilder(commands);
+        Process process = pb.start();
+        boolean exited = process.waitFor(10, TimeUnit.SECONDS);
+        if (!exited) {
+            throw new IllegalStateException("Was not able to wait for ciphertest.sh result");
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-integrationtest/src/main/java/com/daimler/sechub/integrationtest/SecurityTestHelper.java:318]: yes... it was copy waste failure.

[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-integrationtest/src/main/java/com/daimler/sechub/integrationtest/SecurityTestHelper.java:318]: Done + additionally created static variables to define the name only one time

File: sechub-integrationtest/src/main/java/com/daimler/sechub/integrationtest/SecurityTestHelper.java [hunk 4]
```diff
@@ -304,10 +304,29 @@ private String collectionOfSupportedCiphers() throws Exception {
...
+        }
+        int exitCode = process.exitValue();
+        if (exitCode != 0) {
+            throw new IllegalStateException("Was not able to execut curl, exit code was:"+exitCode);
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-integrationtest/src/main/java/com/daimler/sechub/integrationtest/SecurityTestHelper.java:322]: done

File: sechub-integrationtest/src/main/java/com/daimler/sechub/integrationtest/api/AssertSecurityLog.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    private AssertSecurityLog() {
+        logs = TestAPI.getSecurityLogs();
+    }
+
+    public AssertSecurityLog hasEntries(int expectedAmount) {
+        int amount = logs.size();
+        if (expectedAmount!=amount) {
+            // we use equals to have IDE compare also available in outputs
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-integrationtest/src/main/java/com/daimler/sechub/integrationtest/api/AssertSecurityLog.java:28]: The comment was to explain why the asserEquals was used and not a simple fail when already tested. 
But I will remove the comment.

File: sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/AuthorizeValueObfuscator.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+     * After this 
+     * 
+     * @param value
+     * @return
+     */
+    public String obfuscate(String value) {
+
+        int showRealChars = calculateCharsToShow(value);
+        
+        boolean atleastHalfOfThePwdWouldBeObfuscated = value.length() > showRealChars + SHOWN_REAL_CHARS_OF_PASSWORD;
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/AuthorizeValueObfuscator.java:24]: As we discussed before - we will change the obfuscation behaviour and I will remove this variable at all

File: sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/AuthorizeValueObfuscator.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+@Component
+public class AuthorizeValueObfuscator {
+
+    private static final int SHOWN_REAL_CHARS_OF_PASSWORD = 4;
+    private static final String BASIC_WITH_SPACE_LOWERCASED = "basic ";
+
+    /**
+     * Obfuscates given authentication value (password, api token etc.).
+     * If a password starts with "Basic " this will always be visible.
+     * After this
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/AuthorizeValueObfuscator.java:15]: done

File: sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/AuthorizeValueObfuscator.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        
+        if (atleastHalfOfThePwdWouldBeObfuscated) {
+            StringBuilder sb = new StringBuilder();
+            sb.append(value.substring(0, showRealChars));
+            for (int i = showRealChars; i < value.length(); i++) {
+                sb.append('*');
+            }
+            return sb.toString();
+        } else {
+            return "*****length was:" + value.length();
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/AuthorizeValueObfuscator.java:34]: Change the ouptut in another way - already changed to a configurable minimum size parameter.

When called for a password obfuscation with 52 and text has a size of 52 the output will be:
"aW***length:52",
When mimum is not reached, the output will be 
***length:51
This is shorter and easier to read.

File: sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/DefaultSecurityLogService.java [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+        StringBuilder sb = new StringBuilder();
+        sb.append(SECURITY);
+        sb.append("\ndata=\n{}");
+        // we add this to the end, so a wrong defined custom log message could never
+        // overwrite security log data
+        sb.append("\n, message=");
+        sb.append(logData.message);
+
+        /*
+         * convert this to a new list, otherwise slf4j becomes problems with identifying
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/DefaultSecurityLogService.java:108]: OMG: "I become a schnitzel! :facepalm: Thanx a lot!

File: sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/DefaultSecurityLogService.java [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+         * this as list and having wrong output
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/DefaultSecurityLogService.java:109]: done

File: sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/DefaultSecurityLogService.java [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+    // package private so mockable in unit tests
+    Logger getLogger() {
+        return LOG;
+    }
+
+    private SecurityLogData buildLogData(HttpServletRequest request, String sessionId, SecurityLogType type, String message, Object... objects) {
+        SecurityLogData logData = new SecurityLogData();
+        if (type == null) {
+            getLogger().warn("Security log service was called with no log type, so call was wrong implemented! Use fallback:{}", logData.type);
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/DefaultSecurityLogService.java:137]: done

File: sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/DefaultSecurityLogService.java [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+            Map<String, String> headers = logContext.getHttpHeaders();
+            String sanitzedHeaderName = logSanititzer.sanitize(headerName, 40);
+            String sanitizedHeaderValue = logSanititzer.sanitize(headerValue, 1024, false); // headerValue without log forgery handling, we want the origin output
+                                                                                      // inside JSON
+            headers.put(sanitzedHeaderName, sanitizedHeaderValue);
+
+        }
+        if (amountOfHeadersFound > MAXIMUM_HEADER_AMOUNT_TO_SHOW) {
+            getLogger().warn(
+                    "Maximum header values ({}) reached in request from ip={} - header amoutn was {} at all! So truncate header values inside next log entry!",
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/DefaultSecurityLogService.java:195]: done

File: sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/LogSanitizer.java [hunk 13]
```diff
@@ -8,34 +8,51 @@
...
+    public String sanitize(String maybeContaminated, int maxLength) {
+        return sanitize(maybeContaminated, maxLength, true);
+    }
+
+    /**
+     * Returns sanitized text.
+     *
+     * @param maybeContaminated
+     * @param maxLength         when >0 the returned string has this maximum length.
+     *                          use <=0 when Length doesn't matter
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/LogSanitizer.java:39]: done

File: sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/LogSanitizer.java [hunk 14]
```diff
@@ -8,34 +8,51 @@
...
+        }
+        return sanitize(objAsString, maxLength);
+    }
+
+    /**
+     * Returns sanitized text.
+     *
+     * @param maybeContaminated
+     * @param maxLength         when >0 the returned string has this maximum length.
+     *                          use <=0 when Length doesn't matter
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/LogSanitizer.java:26]: done

File: sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/LogSanitizer.java [hunk 15]
```diff
@@ -8,34 +8,51 @@
...
+        return sanitize(maybeContaminated, maxLength, true);
+    }
+
+    /**
+     * Returns sanitized text.
+     *
+     * @param maybeContaminated
+     * @param maxLength         when >0 the returned string has this maximum length.
+     *                          use <=0 when Length doesn't matter
+     * @param handleLogForgery  enables log forgery handling
```
[the remark this answers is not in the document]
Comment (de-jcup) [reply sechub-shared-kernel/src/main/java/com/daimler/sechub/sharedkernel/logging/LogSanitizer.java:40]: Even when this is not introduced inside the PR, it is a good Idea to add the mentioned links.
The sanitizer handles https://cwe.mitre.org/data/definitions/93.html
and https://cwe.mitre.org/data/definitions/117.html

We use slf4 for logging, which should IMHO already handle CWE-116 already.

I will add links for CWE 93 and CWE 117.

**Comment by de-jcup** [comment]
Hm... Analyzing tests did show up that our Junit5 tests are all executed by gradle except in `sechub-integrationtest` sub module!

Must be fixed!

**Review by Jeeppler — APPROVED** [review]

**Pull request merged**